helix-core/src/main/java/org/apache/helix/spectator/RoutingTableProvider.java

```diff
@@ -558,23 +558,28 @@ protected void refresh(Map<String, Map<String, Map<String, CurrentState>>> curre
 
   private void resetRoutingTableAndNotify(long startTime, RoutingTable newRoutingTable) {
     _routingTableRef.set(newRoutingTable);
-    logger.info("Refresh the RoutingTable for cluster {}, takes {} ms.",
-        (_helixManager != null ? _helixManager.getClusterName() : null),
+    String clusterName = _helixManager != null ? _helixManager.getClusterName() : null;
+    logger.info("Refresh the RoutingTable for cluster {}, takes {} ms.", clusterName,
         (System.currentTimeMillis() - startTime));
-    notifyRoutingTableChange();
+
+    notifyRoutingTableChange(_helixManager != null ? _helixManager.getClusterName() : null);
```

Comment:
It's a redundant code. Also, passing null value is risky if it's not handled properly, how about skipping the notify method when it's null?

Reply:
We cannot skip it even it is null. How about let us use Optional<String>?

Reply:
Is it because you'd like to link the first log info with the following logs? 
Optional is a good idea but if it's the purpose of logging, why cannot we move the 1st log into the private method.
Then you can skip if it's null

Reply:
notifyRoutingTableChange() is actually calling the user side registered callback function. The time spent here should be separately accounted for as this is out of our control. This can be potentially moved to some other place, for example in a separate thread specialized for this. That is the reason why I prefer to keep the interface separate, thus log separate.

Reply:
Do I understand it right? The `onRoutingTableChange` at line 579 is an external client call. My suggestion is putting the 
```
logger.info("Refresh the RoutingTable for cluster {}, takes {} ms.", clusterName,
        (System.currentTimeMillis() - startTime));
```
in the first line of `private void notifyRoutingTableChange(String clusterName)`, does it conflict the goal?

Reply:
I think you mean write the following code. Thus, notifyRoutingTableChange() does not need to take any parameter, right?
```
private void notifyRoutingTableChange() 
    notifyRoutingTableChange(_helixManager != null ? _helixManager.getClusterName() : null);
    logger.info("Refresh the RoutingTable for cluster {}, takes {} ms.", clusterName,
        (System.currentTimeMillis() - startTime));
    for (Map.Entry<RoutingTableChangeListener, ListenerContext> entry : _routingTableChangeListenerMap.entrySet()) {
...
```

The thing is that if later you move this notifyRoutingTableChange() to another thread. The logging for "Refresh the RoutingTable for Cluster {} .." time is not accurate, right?

Reply:
If the thread is started immediately, the time log is still correct. If it is a delayed task thread, then it is not accurate. By looking at the code, I think it's unlikely delayed. Isn't it? 
My main concern is still passing NULL to the method, if you have a more reasonable alternative, it's fine. BTW, could you add a TODO to remind the future thread work

Reply:
Will add the TODO. 

Realistically, the database guys showed up 1.3sec delay already. We may indeed to make this callback logic if they complains about it later. Put it another way, the TODO is likely to happen down the line.